torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1280,15 +1300,9 @@ void ProcessGroupNCCL::workCleanupLoop() {
 
             if (dumpOnTimeout_) {
               // Store debug info to storage. (By default to local disk)
-              if (dumpingDebugInfo && dumpingDebugInfo->joinable()) {
-                std::this_thread::sleep_for(
-                    std::chrono::milliseconds(kWatchdogThreadSleepMillis * 30));
-                // At this point, we either have already waited for
-                // `kWatchdogThreadSleepMillis * 30` or the thread has finished
-                // so that we mark the thread detach and the dump of debug info
-                // becomes "best effort".
-                dumpingDebugInfo->detach();
-              }
+              auto& dumpingDebugInfo = tryWriteDebugInfo();
```

Comment:
a little annoying, i couldn't stash the future above since its not copyable and i just took advantage of how tryWrite maintains a static.

If we don't want to keep it that way i can stash a pointer here to avoid multiple calls.

Reply:
I think the lifetime of the return for tryWriteDebugInfo is a bit odd. Can we have it return a future, and then hold an optional future here if we need to have it not exist in certain cases?

Reply:
yep, will do that